File: csharp/bank-account/BankAccount.cs
using System;

public class BankAccount
{
    private bool isOpen = false;
    private float balance;
    private readonly object bankAccountLock = new Object();

    public void Open()
    {
        isOpen = true;
    }

    public void Close()
    {
        isOpen = false;
    }

    public float Balance
    {
        get
        {
            if (!isOpen) throw new InvalidOperationException();

            return balance;
        }
        private set
        {
            if (!isOpen) throw new InvalidOperationException();

            balance = value;
        }
    }

    public void UpdateBalance(float change)
    {
        lock (bankAccountLock)
        {
            Balance += change;
        }
    }
}

File: fsharp/bank-account/BankAccount.fs
module BankAccount

type Account = { mutable balance : decimal ; isOpen : bool }

let mkBankAccount() = { balance = 0.0m; isOpen = false }

let openAccount account = { account with isOpen = true }

let closeAccount account = { account with isOpen = false }

let getBalance account = 
    match account.isOpen with
    | true -> Some account.balance
    | false -> None

let accountLock = new System.Object()

let updateBalance (change: decimal) account =
    lock accountLock (fun _ -> account.balance <- account.balance + change)
    account